src/RealFourierArray.jl

```diff
@@ -0,0 +1,91 @@
+# Lazy structure for an array whose real and fourier part can be accessed by A.real and A.fourier
+# For normalization conventions, see PlaneWaveBasis.jl
+
+import Base.getproperty, Base.setproperty!, Base.fieldnames
+import Base.*, Base.+, Base./
+import Base.eltype
+
+## these functions are minimally optimized for the moment
+
+## TODO this always allocates both real and fourier parts, which might be a bit wasteful
+## TODO implement broadcasting. This seems to be non-trivial, so I just implement a few simple methods
+## TODO implement compressed representation in Fourier space of real arrays
+
+# Tr is the underlying real type
+# T is the type of the array in real space
+# TODO make this concrete by parametrizing on the type of the arrays
+mutable struct RealFourierArray{Tr <: Real, T <: Union{Tr, Complex{Tr}}}
+    basis::PlaneWaveBasis{Tr}
+    _real::AbstractArray{T, 3}
+    _fourier::AbstractArray{Complex{Tr}, 3}
+    _state::Int # 0: only _real set. 1: only _fourier set. 2: both set
+end
+# Type of the real part
+Base.eltype(A::RealFourierArray{Tr, T}) where {Tr, T} = T
+
+# Constructors
+RealFourierArray(basis, real, fourier) = RealFourierArray(basis, real, fourier, 2)
+function RealFourierArray(basis::PlaneWaveBasis{T}; iscomplex=false) where {T}
+    from_real(basis, zeros(iscomplex ? complex(T) : T, basis.fft_size...))
+end
+
+function from_real(basis, real_part::AbstractArray{T}) where {T <: Number}
```

Comment:
I debated that for a while, but ultimately I think `from_real/from_fourier` is clear enough. Somebody confused over what that does will find it easily enough, and there's no possible confusion. Plus, leaving it like that opens the possibility to doing something like `from_real(kpt, real_part)` later on, which could be nice.

Reply:
Sure, but `rfa_from_real(kpt, real_part)` works as well, doesn't it?

Reply:
But then it might be a different kind of realfourierarray

Reply:
Not sure I understand correctly: Because than it only transforms between (full) real-space grid and the ball-like Fourier basis on the k-Point? That's a fair point. In that case I am not too sure whether this mechanism does not end up hiding the lossy conversion real -> Fourier.

But ok, most of the time we will use this interface in such cases
```julia
ρ = from_real(data)
```
where it is implicitly clear from the lhs, what kind of object the `from_real` returns. Not sure that is such a good design, because it relies on the fact that one does not abuse the function like
```julia
from_fourier(my_nasty_function(from_real(data)))
```
which becomes kind of strange, but arguably that's quite constructed.

Reply:
I don't know if we should have a RFA-type thing for each kpoint. Maybe, maybe not. We'll see later on.

I don't understand your problem with `from_fourier(my_nasty_function(from_real(data)))`?

Reply:
I tried `rfa_from_real` but it really looks weird. So let's keep `from_real/from_fourier` for now. The intent is to use these a lot more, eg by having constructors with functions, and by storing potentials in that format, so it'll become common enough for this kind of arrays that we don't have to qualify it each time.

Reply:
Agreed.